spampd.pl

```diff
@@ -983,39 +1062,13 @@ sub process_message {
     my $previous_alarm = alarm($prop->{satimeout});
 
     # Audit the message
-    if ($prop->{sa_version} >= 3) {
-      $mail = $assassin->parse(\@msglines, 0);
-      undef @msglines;  #clear some memory-- this screws up SA < v3
-    }
-    elsif ($prop->{sa_version} >= 2.70) {
-      $mail = Mail::SpamAssassin::MsgParser->parse(\@msglines);
-    }
-    else {
-      $mail = Mail::SpamAssassin::NoMailAudit->new(data => \@msglines);
-    }
-
-    # Check spamminess (returns Mail::SpamAssassin:PerMsgStatus object)
-    my $status = $assassin->check($mail);
-
+    my $status = $self->audit(@msglines);
+    undef @msglines;
```

Comment:
As now the @msglines are not passed as reference it's safe to set it undef here

Reply:
I didn't notice that before. My Perl is rusty but this means `audit()` is making a copy of the lines array (in `my ($self, $msglines) = @_;`), doesn't it? In fact it has to reconstruct the array from individual members (since what is actually passed is `audit($msglines[0], $msglines[1], ...)`.

Reply:
This is fixed now. instead of passing array of message lines. Now i am passing message as string.

Reply:
Well, now SA has to re-split the message into lines.
SA::parse() ends up here: https://github.com/apache/spamassassin/blob/trunk/lib/Mail/SpamAssassin/Message.pm#L184

What was the issue with passing the array of lines by reference?

We still have the issue of re-writing the message using SA each time even if it's not spam or `tag-all`.

My number one priority for these changes is that it doesn't impact existing implementations in any significant way.  All this text parsing is already super expensive, no further overhead should be introduced. If anything I'd love to find ways to make this _more_ efficient.